### ISSUE TITLE
Event showing as all day event but in actual it is not all day event

### ISSUE THREAD

**Issue by anjalikalsariya** [jquense/react-big-calendar#2687]
### Check that this is really a bug

- [X] I confirm

### Reproduction link

https://codesandbox.io/p/sandbox/priceless-morning-9jzt7q?file=%2Fevents.js%3A6%2C31

### Bug description

When event have same start and end date with time 12:00 AM. It is showing as all day event but it should be normal event. That Same event showing as normal event in google and outlook calendar.
 
_Example ,_
Start date is  27 Nov 12:00AM
End date is 27 Nov 12:00AM

_React big calendar that I am usinng_
<img width="1156" alt="image" src="https://github.com/user-attachments/assets/aadb5048-dd84-4c32-9843-bcd6d36dc836">

_Outlook calendar_
<img width="1072" alt="image" src="https://github.com/user-attachments/assets/9406327c-9a31-4803-a6a6-2f809b54746d">

### Expected Behavior

I think it should behave same like outlook and google calendar

Start date is  27 Nov 12:00AM
End date is 27 Nov 12:00AM

### Actual Behavior

Event showing as all day event but in actual it is not all day event

For example of this type of dates

### react-big-calendar version

^1.13.1

### React version

18.3.1

### Platform/Target and Browser Versions

Mac OS -  Chrome

### Validations

- [X] Read the [docs](http://jquense.github.io/react-big-calendar/examples).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.
- [X] Make sure this is a react-big-calendar issue and not an implementation issue

### Would you like to open a PR for this bug?

- [ ] I'm willing to open a PR

**Comment by anjalininjadev** [comment]
Any one can help me please ?

**Comment by dBianchii** [comment]
+1 for me. Would like to see this fixed

**Comment by wcbianchi** [comment]
Please, it will help a lot.

**Comment by nickcarnival** [comment]
Have you guys tried the [`allDayAccessor`](https://jquense.github.io/react-big-calendar/examples/index.html?path=/docs/props--all-day-accessor) ? 

This allows you to define when an event is or is not an all day event

**Comment by anjalikalsariya** [comment]
@nickcarnival  No I didn't try that before. I will try and let you know

**Comment by nickcarnival** [comment]
@anjalikalsariya So I played around a bit with the `allDayAccessor` and was NOT able to resolve the same start and end time showing as allDay issue.

I did find a work around that is not a great fix but _may_ help you out...

You can add one second to the end accessor like so:


```ts
      endAccessor={(e) => {
        return moment(e.end).add(1, "s").toDate();
      }}
```

You'll also want to remove the `allDay: true` from the `events.js`:

![Screenshot 2024-12-26 at 9 35 39 AM](https://github.com/user-attachments/assets/c5076b09-dbf8-4fd8-be6b-16cb9a6d2a70)

This event will no longer show as all day:

![Screenshot 2024-12-26 at 9 34 58 AM](https://github.com/user-attachments/assets/e368d57a-d4f9-45aa-a8ff-632e83107ca7)

**Comment by nickcarnival** [comment]
Apologies for the brain dump but I became curious why the above hack "fixes" the issue. 

TLDR; the all day accessor will be ignored you have to make sure the hour, minute, second or millisecond are different for this event to not be considered an all day event. 

Also, this could be wrong I'm not very familiar with this codebase. 

It does not seem like exclusively altering the `allDayAccessor` will prevent same events that have the same start and end date (including the ms, s, m, h) will from entering the all day event array. 

I believe this is what determines whether or not to append an event to the all days array:
[source code permalink](https://github.com/jquense/react-big-calendar/blob/1a6a9938a7c4122398b44b9c06c4bcfbfec0894d/src/TimeGrid.js#L326)

```ts
    events.forEach((event) => {
      if (inRange(event, start, end, accessors, localizer)) {
        let eStart = accessors.start(event),
          eEnd = accessors.end(event)

        if (
          accessors.allDay(event) ||
          localizer.startAndEndAreDateOnly(eStart, eEnd) ||
          (!showMultiDayTimes && !localizer.isSameDate(eStart, eEnd))
        ) {
          allDayEvents.push(event)
        } else {
          rangeEvents.push(event)
        }
      }
    })
```

Using this date, which I took from your codesandbox, as an example:

```js
const exampleEvent = {
  title: "Event title",
  start: new Date(2015, 3, 0),
  end: new Date(2015, 3, 0)
}
```

We need this block to return false in order to not have an all day event:
```ts
accessors.allDay(event) ||
localizer.startAndEndAreDateOnly(eStart, eEnd) ||
(!showMultiDayTimes && !localizer.isSameDate(eStart, eEnd))
```

Assuming we set our assessor to

```tsx
allDayAccessor={(e) => false}
```

We still have the other conditions to deal with. 

We can set the `showMultiDayTimes` prop to True and skip the `localizer.isSameDate` check. 

This leaves us with only this condition `localizer.startAndEndAreDateOnly(eStart, eEnd)` runs this on both start and end:

```ts
export function isJustDate(date) {
  return (
    dates.hours(date) === 0 &&
    dates.minutes(date) === 0 &&
    dates.seconds(date) === 0 &&
    dates.milliseconds(date) === 0
  )
}
```

So our `exampleEvent` `startAndEndAreAreDateOnly` will return true which means this event is an all day event :)

It seems like the simplest way around this is to just add a second or millisecond to the end date which doesn't seem ideal to me. 

You could probably make a PR that changes this condition:

```ts
accessors.allDay(event) ||
localizer.startAndEndAreDateOnly(eStart, eEnd) ||
(!showMultiDayTimes && !localizer.isSameDate(eStart, eEnd))
```

I'm not sure what the intended behavior would be, right now it seems as if start and end being equivalent SHOULD be an all day event.

**Comment by anjalikalsariya** [comment]
Thanks @nickcarnival,  I tried this solution and its working for me

**- In events array**

<img width="599" alt="image" src="https://github.com/user-attachments/assets/394fad57-655c-4445-9600-ca31a509f5df" />






- **In react big calendar**

```
endAccessor={(event: IEvent) =>

            !event.allDay
              ? new Date((event?.end as Date).getTime() - 1)
              : new Date((event?.end as Date).getTime())
}

```
